### PR TITLE
[proposal] extension to allow eager initialize app scoped beans

### PR DESCRIPTION
**Pull request by Turini**
please note tha this is just a draft... if users and devs like it, I can write some docs and a
few tests (hella don't know how!). This feature was asked a few times at vraptor's user list.

With this PR, VRaptor users may add `@EagerInitialized` annotation at the application 
scoped beans that should be initialized on startup, instead of the first request. In example:

```
@ApplicationScoped @EagerInitialized
public class DummyAppScoped {

	@PostConstruct public void init() {
		System.out.println("DummyAppScoped initialized");
	}
}
```

For now the extension is disabled by default, to use it you need to manually register the
`EagerApplicationScopedExtension` on `services/javax.enterprise.inject.spi.Extension`.
If most of you agree, it can be enabled by default (I don't like it). So... what do you think?

### CODE REVIEW

**Comment by lucascs** [comment]
Isn't it the same as the `@Startup` from CDI?

**Comment by garcia-jj** [comment]
@lucascs `Startup` annotation is from EJB spec. So I don't works when the application runs under a servlet environment.

**Comment by lucascs** [comment]
ops o.o

File: vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		
+		Set<Bean<?>> beans = beanManager.getBeans(Object.class, 
+				new AnnotationLiteral<EagerInitialized>() {});
+		
+		for (Bean<?> bean : beans) {
+			Class<?> beanClass = bean.getBeanClass();
+			if(beanClass.isAnnotationPresent(ApplicationScoped.class)){
+				CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
+				beanManager.getReference(bean, beanClass, ctx).toString();
+			}
```
Comment (lucascs) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: we should error if the bean is not `@ApplicationScoped`

File: vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	public void initialize(@Observes AfterDeploymentValidation event, BeanManager beanManager) {
+		
+		Set<Bean<?>> beans = beanManager.getBeans(Object.class, 
+				new AnnotationLiteral<EagerInitialized>() {});
+		
+		for (Bean<?> bean : beans) {
+			Class<?> beanClass = bean.getBeanClass();
+			if(beanClass.isAnnotationPresent(ApplicationScoped.class)){
+				CreationalContext<?> ctx = beanManager.createCreationalContext(bean);
+				beanManager.getReference(bean, beanClass, ctx).toString();
```
Comment (lucascs) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: don't we have a cdi method to initialize the class? toString on arbitrary classes is not exactly good. #gambiarra

Comment (asouza) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: I think we should use the @Startup annotation. It has a well defined
meaning... Just add the ejb-spec jar and it is done. I don't think it is a
good practice to create annotations that already exists.

Em ter, 3 de fev de 2015 20:26, Lucas Cavalcanti <notifications@github.com>
escreveu:

Comment (asouza) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Besides that, I like the idea :).

Em ter, 3 de fev de 2015 23:43, Alberto SOUZA <alots.ssa@gmail.com>
escreveu:

Comment (garcia-jj) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: @asouza there is a way to break applications with EJB? May be we can do some tests to check this.

Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: @asouza I'm not sure if I get your point. The idea is just to reuse `@Startup` annotation? If so, it would be really confusing and frustrating, since we'd add ejb-api and all its annotations/interfaces only to reuse a single annotation. It'll allow users adding `@DependsOn` trying to order its startup beans, or even use `@Stateless` and so on (that will fail quietly, since it's not implemented). It'd also allow users to use wrong annotations on interceptors, just like `javax.ejb.aroundInvoke`, etc. IMO it doesn't worst. 

If I got the wrong idea, could you elaborate?

Comment (asouza) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: You got it right. If you use vraptor in a application server, you have all
these possibilites, already. So, I don't see as problem use the same
annotations in other container. Finally, I really believe that our focus
should be the application servers, that is why we are using CDI as our DI
container.

Em qua, 4 de fev de 2015 08:18, Rodrigo Turini <notifications@github.com>
escreveu:

In vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/

Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: the point is that using EJB annotations, just like `@DependsOn`, in an app server will 
work. That's not the case here. We'll provide a bunch of annotations that doesn't 
work for servlet container. The goal of this PR is help this kind of user. After all, we 
don't need this `Extension` for application server... 

> Finally, I really believe that our focus should be the application servers

Sure, but we shouldn't forget about non app server users... there's a lot of them =)

Comment (asouza) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Ok, then. We have different perspectives about this subject :). For me, the
happy path is to delete VRaptor's annotations in order to use annotations
from the specs.

Em Wed Feb 04 2015 at 10:56:30 AM, Rodrigo Turini <notifications@github.com>
escreveu:

Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: I'd entirely agree with you if we didn't have so many servlet container users :)
(including us, in our projects)

Comment (asouza) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Yes, we have to evolve :).

Em Wed Feb 04 2015 at 11:08:34 AM, Rodrigo Turini <notifications@github.com>
escreveu:

Comment (csokol) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Can't we use the annotation from the spec but only process it in case
vraptor detects that is not running in a app server? At least we won't need
to create a new annotation.

Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: To use `@Startup` we'll need to add ejb spec as dependency, that will provide a 
bunch of annotations (that wont work on servlet container) and not only `@Startup`.

I've mentioned the problems about doing it [here](https://github.com/caelum/vraptor4/pull/937/files#r24073789).

Comment (csokol) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: :disappointed:

File: vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java (hunk 1 above)
Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: ops, totally! It's terrible to fail quiet.
Do you think it's better to log it (error) or to send an exception?

Comment (lucascs) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Exception

Comment (lucascs) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Exception so the user discover it before having a chance to put it in production.

Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/extension/EagerApplicationScopedExtension.java]: Indeed. I'll change it asap

**Comment by garcia-jj** [comment]
I dislike this feature because CDI beans are lazy for a good reason. The feature to instantiate a bean only when need is pretty cool. But I don't have any problem if this component exists, but not enabled by default.

Another point is: since this is unusual feature, why not create a plugin? I believe that less than 1% users will use this feature (my opinion only, of course). And the adoption of the plugin can be measure the usage, so we can merge into core.

**Comment by asouza** [comment]
Otavio ++.

Em Thu Feb 05 2015 at 4:03:08 PM, Otávio Garcia <notifications@github.com>
escreveu:

**Comment by Turini** [comment]
> The feature to instantiate a bean only when need is pretty cool.

totally, but sometimes, we want to run some code before first request.
If not, there is no reason to EJB `@Startup` exists too :)
When you use app servers, and need it, just use `@Startup` and go be
happy. When you use servlet container, and need it,... Oops

>  if this component exists, but not enabled by default.

it's not enabled by default... it'll cause no difference for who don't want

> since this is unusual feature, why not create a plugin?

why unusual? since the beginning of vraptor 4 users ask about that...
I always incentive the idea of add features first on a plugin, that's easier
to evolve, and it's use can be mensured. But we're talk about an extension.
It's like 10 lines and 1 annotation, disabled by default... for me it doesn't
worst the cost of create a new repository (that will never change) and the
work of release it. If that's the case, we can create a cookbook teaching
users to impl it themselves, not a plugin. But lets wait more opinions here.

TL;DR: too simple and convinient for servlet container users, why not? ;)

what do you think @lucascs @csokol @felipeweb @nykolaslima `@all`?

**Comment by rponte** [comment]
I like the idea. It's simple and useful, so I don't have to implement my
own solution with events.

On Friday, February 6, 2015, Rodrigo Turini <notifications@github.com>
wrote:

> The feature to instantiate a bean only when need is pretty cool.
>
> totally, but sometimes, we want to run some code before first request.
> If not, there is no reason to EJB @Startup exists too :)
> When you use app servers, and need it, just use @Startup and go be
> happy. When you use servlet container, and need it,... Oops
>
> if this component exists, but not enabled by default.
>
> it's not enabled by default... it'll cause no difference for who don't want
>
> since this is unusual feature, why not create a plugin?
>
> why unusual? since the beginning of vraptor 4 users ask about that...
> I always incentive the idea of add features first on a plugin, that's
> easier
> to evolve, and it's use can be mensured. But we're talk about an extension.
> It's like 10 lines and 1 annotation, disabled by default... for me it
> doesn't
> worst the cost of create a new repository (that will never change) and the
> work of release it. If that's the case, we can create a cookbook teaching
> users to impl it themselves, not a plugin. But lets wait more opinions
> here.
>
> TL;DR: too simple and convinient for servlet container users, why not? ;)
>
> what do you think @lucascs <https://github.com/lucascs> @csokol
> <https://github.com/csokol> @felipeweb <https://github.com/felipeweb>
> @nykolaslima <https://github.com/nykolaslima> @all?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/vraptor4/pull/937#issuecomment-73219738>.
>


-- 
Rafael Ponte
TriadWorks | Formação Java
http://cursos.triadworks.com.br

**Comment by asouza** [comment]
Is there a @VRaptorInitialize event, there isn't? I forgot about :(

Em sex, 6 de fev de 2015 09:04, Rafael Ponte <notifications@github.com>
escreveu:

**Comment by garcia-jj** [comment]
@Turini EJB have startup events because have features like transactions, scheduling, message queues and so on. And vraptor-core is only a controller in the MVC pattern. There is no view, there is no model. Features that is not about controllers needs to live in its own plugin. This is a reason why jpa, simplemail, and more optional features are plugins.

Feature like converter overriding is more useful than this extension, and was created as plugin. So why this feature is better than overriding converters to live in core? There are more than 4 users that asking for converter overriding.

**Comment by linyatis** [comment]
@Turini, as a user of VRaptor, I agree with @garcia-jj that this is an unusual feature. I think that 4 users is not substantial. I would like it as a plugin.

But, it is a great feature! :smiley:

**Comment by garcia-jj** [comment]
@Turini Only to explain... I'm only discussing technical points about the best place to implement the feature. As @linyatis said, this is a good feature (for users who want to use) and the code you wrote is too good also. The point is only where to implement.

**Comment by Turini** [comment]
I'm sure about that, no worries! This kind of discution is very healthy for the 
project. I'll move this code to a plugin as soon as possible. Thank you guys.

**Comment by lucascs** [comment]
Can we tell the users on the docs that, on AS they can use `@Startup`, and on servlet containers they can:

```java
@ApplicationScope
public class MyEagerClass {
    
     public void init(@Observes VRaptorInitialized init) {
          System.out.println("Do the interesting stuff here");
     }
}
```
?

**Comment by Turini** [comment]
`VRaptorInitialized` event runs on the first request, not startup

**Comment by lucascs** [comment]
What about telling the users to use `@Observes AfterDeploymentValidation event`, or some other CDI event?

**Comment by asouza** [comment]
Maybe this is the solution =>
http://www.next-presso.com/2014/06/you-think-you-know-everything-about-cdi-events-think-again/

There is the @Initialized and it can be used with Observers. Does it worth
a try?

Em Mon Feb 09 2015 at 10:52:46 AM, Lucas Cavalcanti <
notifications@github.com> escreveu:

**Comment by lucascs** [comment]
`public void myInit(@Observes @Initialized(ApplicationScope.class) ServletContext ctx)`

does it happen before the first request?

**Comment by Turini** [comment]
@lucascs About observing `AfterDeploymentValidation`, it's only possible for `Extensions` 
(just like I did on my solution). @asouza for sure we can use `@Initialized`, just like we did
on [ServletContextFactory](https://github.com/caelum/vraptor4/blob/master/vraptor-core/src/main/java/br/com/caelum/vraptor/ioc/cdi/ServletContextFactory.java). But our bean will run just before VRaptor filter (not sure if it will be 
a problem). And it's much more verbose than just adding a simple `@Eager` on the bean, but 
anyway... I think it'd be a solution for those who are using servlet containers. I'm ok with it :)
